### PR TITLE
Unifies pagination template usage (#6531)

### LINKED ISSUE #6531: Unified pagination template
There is a template for the pagination bar (templates/base/paginate.tmpl). However, this template is not used in all templates that use a pagination.

The HTML structure is the same. The only difference is the additional URL params used for the pagination links. The pagination template should be adjusted to be able to configure these URL params and be used on all templates that use pagination.

### PR DESCRIPTION
**Pull request by saitho**
This pull request makes the additional URL params that are used in the pagination template configurable. This enables the usage of the same template among all pages that use a pagination.

Resolves #6531.

---

### How to test
Check the paginations on all pages that use one, i.e. templates with `{{template "base/paginate" .}}`

- General
  - Explore repositories: http://localhost:3000/explore/repos
  - Explore users: http://localhost:3000/explore/users
  - Explore organisations: http://localhost:3000/explore/organizations
  - Explore code (redirects to explore/repos?!): http://localhost:3000/explore/code
- User
  - User home: http://localhost:3000
  - User dashboard Issues/Pulls: http://localhost:3000/issues / http://localhost:3000/pulls
  - User profile: http://localhost:3000/[username]
  - User notifications: http://localhost:3000/notifications
  - Organisation profile: http://localhost:3000/[orgname]
- Repository
  - Repository search: http://localhost:3000/[owner]/[repo]/search
  - Issue List: http://localhost:3000/[owner]/[repo]/issues
  - Milestones: http://localhost:3000/[owner]/[repo]/milestones
  - Releases: http://localhost:3000/[owner]/[repo]/releases
  - Milestone details: http://localhost:3000/[owner]/[repo]/milestone/[milestone]
  - template repo/commits_table, e.g. repository commits: http://localhost:3000/[owner]/[repo]/commits/branch/master
- Admin
  - Admin: organisation list: http://localhost:3000/admin/orgs
  - Admin: user list: http://localhost:3000/admin/users
  - Admin: repository list: http://localhost:3000/admin/repos

### CODE REVIEW

File: routers/admin/notice.go [hunk 1]
```diff
@@ -5,14 +5,12 @@
...
 import (
-	"github.com/Unknwon/com"
-	"github.com/Unknwon/paginater"
-
 	"code.gitea.io/gitea/models"
 	"code.gitea.io/gitea/modules/base"
 	"code.gitea.io/gitea/modules/context"
 	"code.gitea.io/gitea/modules/log"
 	"code.gitea.io/gitea/modules/setting"
+	"github.com/Unknwon/com"
```
Comment (lunny) [inline routers/admin/notice.go:15]: blank line

File: modules/context/pagination.go [hunk 2]
```diff
@@ -2,27 +2,36 @@ package context
...
+// NewPagination creates a new instance of the Pagination struct
+func NewPagination(total int, page int, issueNum int, numPages int) *Pagination {
+	p := &Pagination{}
+	p.Paginater = paginater.New(total, page, issueNum, numPages)
+	p.PaginationLinkParams = ""
+	return p
+}
+
+// AddPaginationParam adds a value from context identified by ctxKey as link param under a given paramKey
+func (p *Pagination) AddPaginationParam(ctx *Context, paramKey string, ctxKey string) {
```
Comment (lunny) [inline modules/context/pagination.go]: Since it's a method of struct now, the name could be `AddParam`

File: modules/context/pagination.go [hunk 3]
```diff
@@ -2,27 +2,36 @@ package context
...
+	p.PaginationLinkParams = template.URL(fmt.Sprintf("%v%s=%v&", p.PaginationLinkParams, paramKey, ctx.Data[ctxKey]))
 }
 
-func DefaultPaginationParams(ctx *Context) {
-	ClearPaginationParam(ctx)
-	AddPaginationParam(ctx, "sort", "SortType")
-	AddPaginationParam(ctx, "q", "Keyword")
-	AddPaginationParam(ctx, "tab", "TabName")
+// DefaultPaginationParams provides common pagination params that are often used
+func (p *Pagination) DefaultPaginationParams(ctx *Context) {
```
Comment (lunny) [inline modules/context/pagination.go]: As above, the method name could be `ResetDefaultParams`

**Review by lunny — CHANGES_REQUESTED** [review]

File: modules/context/pagination.go [hunk 4]
```diff
@@ -0,0 +1,37 @@
+package context
```
Comment (techknowlogick) [inline modules/context/pagination.go:5]: Please add the copyright header to this file

File: modules/context/pagination.go [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+import (
+	"fmt"
+	"github.com/Unknwon/paginater"
```
Comment (techknowlogick) [inline modules/context/pagination.go]: please split internal and external packages with an empty line.

File: routers/admin/notice.go [hunk 6]
```diff
@@ -5,14 +5,12 @@
 package admin
```
Comment (techknowlogick) [inline routers/admin/notice.go:6]: Please add Gitea to copyright header

File: routers/repo/commit.go [hunk 7]
```diff
@@ -14,8 +14,6 @@ import (
 	"code.gitea.io/gitea/modules/git"
 	"code.gitea.io/gitea/modules/log"
```
Comment (techknowlogick) [inline routers/repo/commit.go:16]: Please add copyright header to this file

File: routers/user/notification.go [hunk 8]
```diff
@@ -6,8 +6,6 @@ import (
 	"strconv"
 	"strings"
 
-	"github.com/Unknwon/paginater"
```
Comment (techknowlogick) [inline routers/user/notification.go:9]: please add copyright header to this file

File: modules/context/pagination.go [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+	return p
+}
+
+// AddParam adds a value from context identified by ctxKey as link param under a given paramKey
+func (p *Pagination) AddParam(ctx *Context, paramKey string, ctxKey string) {
+	_, exists := ctx.Data[ctxKey]
+	if !exists {
+		return
+	}
+	p.paginationLinkParams = template.URL(fmt.Sprintf("%v%s=%v&", p.paginationLinkParams, paramKey, ctx.Data[ctxKey]))
```
Comment (zeripath) [inline modules/context/pagination.go]: I'm fairly certain you need to be escaping some of those. 

```suggestion
	p.paginationLinkParams = template.URL(fmt.Sprintf("%v%s=%v&", p.paginationLinkParams, url.QueryEscape(paramKey), url.QueryEscape(ctx.Data[ctxKey])))
```

Comment (saitho) [reply modules/context/pagination.go]: Wouldn't do any harm :) Can you suggest the required import _net/url_ as well? Then I can commit it via GitHub. ^^

File: modules/context/pagination.go [hunk 10]
```diff
@@ -0,0 +1,47 @@
+// Copyright 2019 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package context
+
+import (
+	"fmt"
+	"html/template"
```
Comment (zeripath) [inline modules/context/pagination.go]:
```suggestion
	"html/template"
	"net/url"
```

File: modules/context/pagination.go (hunk 9 above)
Comment (zeripath) [reply modules/context/pagination.go]: Cheeky. :-p

In order to convince you it was needed consider what would have happened had you had a & or # in keyword.

**Review by zeripath — APPROVED** [review]

**Comment by zeripath** [comment]
There's a Conflict with a commit in master.

Otherwise this now looks good

File: modules/context/pagination.go [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+	"net/url"
+
+	"github.com/Unknwon/paginater"
+)
+
+// Pagination provides a pagination via Paginater and additional configurations for the link params used in rendering
+type Pagination struct {
+	Paginater            *paginater.Paginater
+	paginationLinkParams template.URL
```
Comment (lunny) [inline modules/context/pagination.go]: rename `paginationLinkParams` to `linkParams`

File: modules/context/pagination.go [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+	"fmt"
+	"html/template"
+	"net/url"
+
+	"github.com/Unknwon/paginater"
+)
+
+// Pagination provides a pagination via Paginater and additional configurations for the link params used in rendering
+type Pagination struct {
+	Paginater            *paginater.Paginater
```
Comment (zeripath) [inline modules/context/pagination.go]:
```suggestion
	Paginater  *paginater.Paginater
```

**Review by lunny — APPROVED** [review]

**Pull request merged**